content/blog/2025-04-27-valkey-modules-rust-sdk-updates.md

```diff
@@ -0,0 +1,272 @@
+We briefly touched on how to use [Valkey Modules Rust SDK](https://github.com/valkey-io/valkeymodule-rs) to build a simple module.
+In this article we will expand on this topic and discuss some of the new features added to the SDK over the last year.
+SDK itself also provides examples on how to use these new features but this article summarizes them below.
+
+## What is Valkey Modules Rust SDK?  
+
+The SDK is based on [Redis Modules Rust SDK](https://github.com/RedisLabsModules/redismodule-rs) and provides abstraction APIs on top of Valkey Modules own API.
+For those familiar with Rust development the SDK is a regular Rust crate that can be added to Cargo.toml file like any other Rust dependency.
+It requires functionality to be present in the underlying Valkey version but allows writing Valkey modules in Rust, without needing to use raw pointers or unsafe code.
+Recently released [Bloom Filers module](/blog/introducing-bloom-filters/) is built using this crate and several of the developers who worked on the module contributed to the SDK.
+Let's deep dive into to new features.
+
+## Client
+
+`Context` struct now has several new functions to get information about the client.
+It provides Rust wrappers around `Module_GetClient*` functions in the underlying Valkey Module API.
+Most new functions return `ValkeyResult` so the module developer can handle the error appropriately.
+The functions can be called for the current client or by specifying `client_id`.
+
+```rust
+fn my_client_cmd(ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
+    let client_id = ctx.get_client_id();
+    let username = ctx.get_client_username_by_id(client_id);
+    Ok(ValkeyValue::from(username.to_string()))
+}
+valkey_module! {
+    ...
+    commands: [
+        ["my_client_cmd", my_client_cmd, "", 0, 0, 0],
+    ]
+}
+```
+
+## Auth callback
+
+Valkey 7.2 introduced support for callbacks after authentication.
+Now it can be leveraged via modules.
+
+```rust
+fn auth_callback(
+    ctx: &Context,
+    username: ValkeyString,
+    _password: ValkeyString,
+) -> Result<c_int, ValkeyError> {
+    if ctx.authenticate_client_with_acl_user(&username) == Status::Ok {
+        // can be combined with ctx.get_client_ip to control access by IP address
+        let _client_ip = ctx.get_client_ip()?;
+        ...
+        return Ok(AUTH_HANDLED);
+    }
+    Ok(AUTH_NOT_HANDLED)
+}
+valkey_module! {
+    ...
+    auth: [auth_callback],
+}
+```
+
+## Preload validation
+
+`valkey_module!` macro already provided `init` callback that can be used to execute custom code during module load.
+However, `init` executes at the very end of module load after new commands and data types are created.
+That can be use useful but what if we wanted to stop module load before any of that happens?
+For example, we might need to restrict module to be loaded only on specific version of Valkey.
```

Comment:
'we might need to restrict module' `->` 'we might need to restrict a module'

Reply:
updated